Qorus login fp

rishabhmalik759 opened the pull request.

Commit:
Bootstraped project
- Installed required library
- Added tsconfig and jest config
- Updated Readme
- Initialized logger

Commit:
Added qorus login functionality with tests

Commit:
Added login and edited project structure

Commit:
updated readme name

Commit:
Added qorul login functionality along with:
- eslint
- updated build config
- removed logger
- added catchAll decorator

Commit:
Added logger decorator

Commit:
Updated Qorus Integration
- Updated tests
- Added .env configurations

Commit:
removed unwanted console log

Commit:
Added LocalStorageManger
- updated tests
- updated login to use localstorage first with token validation

Commit:
updated linting and localstorage

Commit:
Updated QorusIntegration and local storage manager
- Api version modification added
- Converted local storage manager from class to function
- Added types

Commit:
Added Inheritance with Qorus Instance

Commit:
Added Qorus Authenticator function
- Added support for multiple endpoints
- Updated all the functions to support multiple endpoints
- Updated Tests

Commit:
updated tests and exports from index

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+    return false;
+  };
+
+  const initEndpnt = (props: IInitEndpoint): IEndpoint => {
```
Please don't unnecessarily abbreviate just to save 1 or 2 characters `endpnt` vs `endpoint`. These function names make no sense.

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+    return false;
+  };
+
+  const setEndpntUrl = (props: ISetEndpointUrl): boolean => {
```
If the user sets new endpoint URL, that old endpoint needs to be logged out.

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+  const initEndpnt = (props: IInitEndpoint): IEndpoint => {
+    const { id, url, version = 'latest' } = props;
+    const newEndpnt = createEndpoint({ id, url, version });
+    const index = getEndpntIndxById(id);
```
why not use just `const ep = endpoints.find((endpoint) => endpoint.id === id)`?

Foxhoundn commented on src/utils/localStoreMan.ts:
```diff
@@ -0,0 +1,45 @@
+/**
+ * A helper function to check if the running environment is node or client
```
Please do not abbreviate so much, this file could be in a folder called `Managers / LocalStorage.ts`

Foxhoundn commented on test/QorusIntegration.test.ts:
```diff
@@ -0,0 +1,70 @@
+      expect(QorusAuth.getSelectedEndpnt()?.url).toEqual('https://www.google.com');
+    });
+  });
+});
```
Also add negative tests, and multiple instance tests please.

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+const initialEndpoint: IEndpoint[] = [{ url: '', id: '', version: '' }];
+
+const QorusAuthenticator = (): IQorusAuthenticator => {
+  const endpoints: IEndpoint[] = initialEndpoint;
```
there should be no initial endpoint, set it to `[]`

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+const apiPathsInitial: IApiPaths = {
+  login: `/api/latest/public/login`,
+  logout: `/api/latest/logout`,
+  validateToken: `/api/latest/system?action=validateWsToken`,
```
move this to a new file, where we can store all potential API paths in the future, maybe make it an Enum as well?

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+}
+
+interface ISetEndpointVersion {
+  version: string;
```
You use `version` in multiple interfaces, which means it should be extracted to a type. That means we can also control it (because we know that `type TQorusAPIVersion = 1 | 2 | 3 | 4 | 5 | 6 | 'latest`

rishabhmalik759 commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,286 @@
+    getApiPaths,
+    getAllEndpnts,
+  };
+};
```
We don't want the user to access the endpoint's token parameter, I exposed the getAllEndpoints function to get the current endpoints array

Commit:
Added comments and updated tests

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -76,54 +75,89 @@ const createEndpoint = (props: IEndpoint): IEndpoint => {
+   */
+  const initEndpoint = async (props: IInitEndpoint): Promise<IEndpoint> => {
+    const { id, url, version } = props;
+    const newEndpoint = createEndpoint({ id, url, version: version ? version : 'latest' });
```
Why is createEndpoint not part of this function and why does it not return already the index / endpoint created?

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -172,55 +212,98 @@ const QorusAuthenticator = (): IQorusAuthenticator => {
       }
       endpoints[index].authToken = undefined;
       selectedEndpoint.authToken = undefined;
     }
```
I'd move this to a function `removeEndpoint`

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -172,55 +212,98 @@ const QorusAuthenticator = (): IQorusAuthenticator => {
+   */
+  const getEndpointVersion = (id?: string): Version | undefined => {
+    if (id) {
+      const index = getEndpointIndxById(id);
```
You repeat this a lot, why not simply have a function `getEndpointById` and eliminate all the index searching?

Foxhoundn replied:
```
const maybeGetEndpointById: Endpoint | undefined = (id: string) => endpoints.find((endpoint) => endpoint.id === id);

const maybeGetEndpointData: string | undefined = (id: string, key: string) => {
   return maybeGetEndpointById(id)?.[key];
}

const getEndpointVersion = (id: string) => {
    return maybeGetEndpointData(id, 'version');
}
```

Commit:
Updated Qorus Authenticator
- Added noauth support
- Updated and added tests
- Added axios helpers

Commit:
Fixed documentation issue

Commit:
Added logger and Qorus Request
- Added Winston logger
- Added QorusRequest axios wrapper
- Added Error Handling tests
- Updated QorusAuthenticator

Commit:
updated interfaces

Foxhoundn commented on src/utils/QorusRequest.ts:
```diff
@@ -0,0 +1,64 @@
+};
+
+export default QorusRequest;
+export const QorusReq = QorusRequest();
```
Rename the function to `_QorusRequest` and export `QorusRequest` instead of `QorusReq`

Commit:
fixed setEndpoint url and getEndpointURl

Commit:
updated docs and added CI for tests

Commit:
updated CI for tests

Commit:
updated CI tests

Commit:
updated CI tests

Commit:
Updated docs

Foxhoundn committed:
Code cleaning & export fixes

Commit:
updated comments

Commit:
Added Qorus request to docs

Foxhoundn committed:
Release and publishing updates

Commit:
Merge branch 'qorus-login-fp' of https://github.com/qoretechnologies/qorus-toolkit into qorus-login-fp

Commit:
added docs updated package.json

Commit:
Added shell script to update docs

Commit:
Added scripts to update docs

Commit:
updated lint

Commit:
updated lint

Commit:
updated package.json

Commit:
updated prepush

Commit:
updated version

Commit:
updated package.json

Commit:
updated package.json

Commit:
updated package.json

Commit:
updated updateDocs script

Commit:
updated update docs script

Commit:
Updated the package.json and localStorage manager

Commit:
updated localStorage manager

Commit:
converted library from type module to commonjs

Commit:
updated tsconfig.json

Commit:
updating version

Commit:
Completed runkit comments for QorusAuthenticator

Commit:
Updated updateDocs script

Commit:
Added tests and updated localStorageManager

Commit:
Added TryInRepl button to the docs

Commit:
Fixed lint

Commit:
Updated tests and documentation

Commit:
Updated comments on docs

Commit:
Added return comments in runkit code, updated test

Commit:
Updated redme.md

davidnich commented:
this part: `A authentication module for QoreTechnologies, allows user to create multiple connections as endpoints and manage authentication to them.` should probably read: `An authentication module for Qorus Integration Engine(R), allows users to create multiple connections as endpoints and manage authentication to them.`

Foxhoundn committed:
VIsual changes

Foxhoundn committed:
Remove env file from git ignore

Commit:
Added snapshot tests and coverage tags

Commit:
fixed lint issues

Commit:
update functions script working for one function

Commit:
Added examples for the Authenticator
- addExample script working for all the functions in Authenticator
- got all the

Commit:
Add examples to docs preprocessor completed

Commit:
fixed linting issues

Commit:
Added coverage tags to docs

Commit:
fixed docs for getEndpoinById

Commit:
updating snapshots

Commit:
added snapshots

Commit:
fixed a test

Commit:
removed snapshots

Commit:
updating snapshots

Commit:
updatupdateded snapshots

Commit:
updated tests.yml

Commit:
updated snapshot

Commit:
updated test.yml

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -0,0 +1,505 @@
+
+export interface Authenticator {
+  /**
+   * -login-function takes optional username and password parameters to authenticate the user.
```
We need to remove this from sources when the documentation is built - otherwise it's gonna clutter the in-IDE suggestions and intellisense.

Foxhoundn committed:
Documentation & logout fixes.

Commit:
Added contribution guide
- Updated Readme.md
- Updated QorusAuthenticator comments

Commit:
updated function definition

Commit:
Added cleanCode script and updated CI

Commit:
Updated cleanCode Script

Commit:
updated cleanCode.sh

Commit:
updated cleanCode CI

Commit:
updated cleanCode Script

Commit:
updated version

Commit:
updated docs with interface inheritance

Commit:
Removed unused interface

Commit:
Removed unwanted interface

Commit:
removed unwanted exports

Commit:
Updated export to include missing types

Foxhoundn commented on src/QorusAuthenticator.ts:
```diff
@@ -401,7 +424,7 @@ const _QorusAuthenticator = (): Authenticator => {
-   * A setter to set the url of the selected endpoint
+   * A setter to set the url of the selected {@link Endpoint}
    */
   const setEndpointUrl = async (url: string, id: string = selectedEndpoint.id): Promise<string | undefined> => {
```
This is going to crash if this function is called and `selectedEndpoint` is not defined yet. Change it to `= selectedEndpoint?.id`

Foxhoundn replied:
It's strange that TypeScript did no complain about this. Do we have `strictNullChecks` enabled?

rishabhmalik759 replied:
yes we do have the check in tsconfig

Commit:
Updated setEndpointUrl and setEndpointVersion

Commit:
Updated package.json version to 0.1.2

Commit:
Added new release version

Commit:
Updated release version

Commit:
new version

Commit:
updated CI

Commit:
updated CI

Commit:
updated CI
